src/app/core/tasks/threadpool.nim

```diff
@@ -1,260 +1,59 @@
-import # std libs
-  atomics, json, sequtils, tables
-
 import # vendor libs
-  chronicles, chronos, json_serialization, task_runner
+  json_serialization, json, chronicles, taskpools
 
 import # status-desktop libs
-  ./common
+  ./common 
 
-export
-  chronos, common, json_serialization
+export common, json_serialization
 
 logScope:
   topics = "task-threadpool"
 
 type
   ThreadPool* = ref object
-    chanRecvFromPool: AsyncChannel[ThreadSafeString]
-    chanSendToPool: AsyncChannel[ThreadSafeString]
-    thread: Thread[PoolThreadArg]
-    size: int
-    running*: Atomic[bool]
-  PoolThreadArg = object
-    chanSendToMain: AsyncChannel[ThreadSafeString]
-    chanRecvFromMain: AsyncChannel[ThreadSafeString]
-    size: int
-  TaskThreadArg = object
-    id: int
-    chanRecvFromPool: AsyncChannel[ThreadSafeString]
-    chanSendToPool: AsyncChannel[ThreadSafeString]
-  ThreadNotification = object
-    id: int
-    notice: string
-
-# forward declarations
-proc poolThread(arg: PoolThreadArg) {.thread.}
+    pool: Taskpool
+  ThreadSafeTaskArg* = distinct cstring
 
 const MaxThreadPoolSize = 16
 
-proc init(self: ThreadPool) =
-  self.chanRecvFromPool.open()
-  self.chanSendToPool.open()
-  let arg = PoolThreadArg(
-    chanSendToMain: self.chanRecvFromPool,
-    chanRecvFromMain: self.chanSendToPool,
-    size: self.size
-  )
-  createThread(self.thread, poolThread, arg)
-  # block until we receive "ready"
-  discard $(self.chanRecvFromPool.recvSync())
+proc safe*[T: TaskArg](taskArg: T): ThreadSafeTaskArg =
+  var 
+    strArgs = taskArg.encode()
+    res = cast[cstring](allocShared(strArgs.len + 1))
 
-proc newThreadPool*(size: int = MaxThreadPoolSize): ThreadPool =
-  new(result)
-  result.chanRecvFromPool = newAsyncChannel[ThreadSafeString](-1)
-  result.chanSendToPool = newAsyncChannel[ThreadSafeString](-1)
-  result.thread = Thread[PoolThreadArg]()
-  result.size = size
-  result.running.store(false)
-  result.init()
+  copyMem(res, strArgs.cstring, strArgs.len)
+  res[strArgs.len] = '\0'
+  res.ThreadSafeTaskArg
 
+proc toString*(input: ThreadSafeTaskArg): string =
+  result = $(input.cstring)
+  deallocShared input.cstring
+  
 proc teardown*(self: ThreadPool) =
-  self.running.store(false)
-  self.chanSendToPool.sendSync("shutdown".safe)
-  self.chanRecvFromPool.close()
-  self.chanSendToPool.close()
-  trace "[threadpool] waiting for the control thread to stop"
-  joinThread(self.thread)
-
-proc start*[T: TaskArg](self: Threadpool, arg: T) =
-  self.chanSendToPool.sendSync(arg.encode.safe)
-  self.running.store(true)
-
-proc runner(arg: TaskThreadArg) {.async.} =
-  arg.chanRecvFromPool.open()
-  arg.chanSendToPool.open()
+  self.pool.syncAll() 
+  self.pool.shutdown()
```

Comment:
@alexjba I guess that it's enough to call `shutdown` here (without `syncAll`, more or less in this context they are doing the same thing), cause `teardown` is called on app close only.

But I am wondering, what if we have some task running for some longer time, for example some timer on Nim side, which will trigger an action every 5 mins. Does that mean user will be forced to wait for such task to complete in order the app to gets closed safely? Do we have a way to abort the task execution somehow, when we want,  regardless to the progress of the task being executing?

Reply:
>  Does that mean user will be forced to wait for such task to complete in order the app to gets closed safely?

From what I see yes, it will be blocked on sync. But I'm not sure how we can implement this. Considering the thread is busy running the task, it can't receive any abort signal to cancel the running task.

Reply:
This is a problem with the current threadpool as well. That's why we moved some of the longer async tasks on the status-go side.